### PR TITLE
add client-related middleware

### PR DESCRIPTION
**Pull request by kiurchv**

### CODE REVIEW

File: apps/graphql/lib/graphql_web/middleware/client_authorization.ex [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+      defoverridable middleware: 3
+    end
+  end
+
+  def call(%{state: :unresolved, context: context} = resolution, meta_key: meta_key, context_key: context_key) do
+    allowed_clients = Type.meta(resolution.definition.schema_node, meta_key)
+    current_client = Map.get(context, context_key)
+
+    if Enum.member?(allowed_clients, current_client) do
```
Comment (progsmile) [inline apps/graphql/lib/graphql_web/middleware/client_authorization.ex]: `current_client in allowed_clients` optional

File: apps/graphql/lib/graphql_web/middleware/client_authorization.ex [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      resolution
+    else
+      Resolution.put_result(resolution, {:error, format_forbidden_error()})
+    end
+  end
+
+  def call(res, _), do: res
```
Comment (progsmile) [inline apps/graphql/lib/graphql_web/middleware/client_authorization.ex]: what is `res` - result or resolution? write full name please

File: apps/graphql/lib/graphql_web/middleware/client_metadata.ex [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    with {:ok, headers} <- Map.fetch(context, headers_key),
+         client_id <- get_client_id(headers),
+         {:ok, metadata} <- get_metadata(required_metadata, client_id, headers, opts) do
+      %{resolution | context: Map.merge(context, metadata)}
+    else
+      _ -> Resolution.put_result(resolution, {:error, format_unauthenticated_error()})
+    end
+  end
+
+  def call(res, _), do: res
```
Comment (progsmile) [inline apps/graphql/lib/graphql_web/middleware/client_metadata.ex]: the same

File: apps/graphql/lib/graphql_web/middleware/client_authorization.ex (hunk 2 above)
Comment (kiurchv) [reply apps/graphql/lib/graphql_web/middleware/client_authorization.ex]: The `resolution`, for sure. As defined by `Absinthe.Middleware` behaviour

File: apps/graphql/lib/graphql_web/middleware/helpers.ex [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      extensions: %{code: "UNAUTHENTICATED"}
+    }
+
+  def format_forbidden_error,
+    do: %{
+      message: @forbidden_error_message,
+      extensions: %{code: "FORBIDDEN"}
+    }
+
+  def format_forbidden_error(exception),
```
[the remark this answers is not in the document]
Comment (kiurchv) [reply apps/graphql/lib/graphql_web/middleware/helpers.ex]: Agree

File: apps/graphql/lib/graphql_web/middleware/client_authorization.ex (hunk 1 above)
Comment (kiurchv) [reply apps/graphql/lib/graphql_web/middleware/client_authorization.ex]: What are optional? Please, explain wider

Comment (progsmile) [reply apps/graphql/lib/graphql_web/middleware/client_authorization.ex]: `current_client in allowed_clients` looks more declarative, optional for your fix

**Pull request merged**